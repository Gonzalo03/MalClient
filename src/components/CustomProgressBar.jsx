import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import "../css/animate.css";

export const CustomProgressBar = ({ percent, color, colorAcent, text }) => {
  const [progress, setProgress] = useState(50);

  useEffect(
    () =>
      setTimeout(() => {
        setProgress((percent / 10) * 100 > 100 ? 100 : (percent / 10) * 100);
      }, 50),
    [setProgress, percent]
  );

  return (
    <div
      className={`w-full bg-icon rounded-xl text-xs text-center text-white ${
        progress === 0 && "py-0.5"
      }`}
    >
      <div
        className={`py-0.5 rounded-xl ${
          progress === 0 && "hidden"
        } progress bg-${color}${colorAcent ? `-${colorAcent}` : "-500"}`}
        style={{ width: `${progress <= 10 ? 100 : progress}%` }}
      >
        {`${text} : ${percent}`}
      </div>
      {progress === 0 && `${text} : ${percent}`}
    </div>
  );
};

CustomProgressBar.propTypes = {
  percent: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
};

CustomProgressBar.defaultProps = {
  colorAcent: null,
  percent: 100,
};
