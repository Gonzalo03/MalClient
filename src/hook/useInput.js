import { useState } from "react"

export const useInput = () => {

    const [input, setInput] = useState('');

    const handleInput = (e)=>{

        setInput(e.target.value)

    }

    const resetInput = () => {


        setInput('');

    }

   



    return [input, handleInput, resetInput]

}
