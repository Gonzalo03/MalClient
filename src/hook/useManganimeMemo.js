import { useMemo } from "react";


export const useManganimeMemo = (avgManganime) => {

      const animeMemo = useMemo(() => {
        const avg = avgManganime.reduce((a, b) => a + b, 0);
        return avg / avgManganime.length;
      }, [avgManganime]);
    
    return animeMemo;
}