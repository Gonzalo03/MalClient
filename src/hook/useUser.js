import { useEffect, useState } from "react"
import { request } from "../helpers/request"

export const useUser = (tag) => {
    
    const [user, setUser] = useState({
        loading:true,
        data:null,
        err:null
    })
    
    useEffect(
        ()=>request('user', tag)
        .then(data=>setUser({
                loading:false,
                data,
                err:null
            }))
        .catch(err=>setUser({
                loading:false,
                data:null,
                err
            }))
    , [tag])

        return user;

}
