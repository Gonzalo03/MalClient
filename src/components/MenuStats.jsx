import { CustomProgressBar } from "./CustomProgressBar";

export const MenuStats = ({ stats }) => {
    return (
        <>
        <p className="text-center text-xl text-white"><span className="bg-primary p-2 rounded-xl">{stats.days_watched ? 'Anime Stats' : 'Manga Stats'}</span></p>
        <div className="flex flex-col space-y-3">
            <CustomProgressBar color='gray' percent={stats.days_watched || stats.days_read} text='Spend days'/>
            <CustomProgressBar color='green' percent={stats.watching || stats.reading} text='Current'/>
            <CustomProgressBar color='blue' colorAcent="700" percent={stats.completed} text='Completed'/>
            <CustomProgressBar color='yellow' percent={stats.on_hold} text='On hold'/>
            <CustomProgressBar color='red' percent={stats.dropped} text='Dropped'/>
            <CustomProgressBar color='blue' percent={stats.mean_score} text='Average'/>
        </div>
        </>
    )
}

