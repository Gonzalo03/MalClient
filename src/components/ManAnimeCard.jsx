import { useEffect } from "react";
import PropTypes from 'prop-types'

import { useManime } from "../hook/useManime";

import { CircleLoading } from "react-loadingg";

export const ManAnimeCard = ({ manAnime, context, dispatchManime }) => {
    
  const { loading, data, err } = useManime(manAnime.mal_id, context);
  

  useEffect(
      ()=> data && dispatchManime({type : context, payload : data.score})
  , [dispatchManime, data, context])

  return (
    <>
      {!loading ? (
        <div className="flex rounded overflow-hidden shadow-lg">
          <img className="w-40 object-cover" src={data.image_url} alt={data.title} />
          <div className="px-6 py-4">
            <div className="font-bold text-xl text-custom-first mb-2 w-60">
              {data.title}
            </div>
            <p className="text-gray-700 text-base text-custom-second flex flex-col space-y-2">
              <span>
                Status :
                <span
                  className={`${
                    data.status.includes('Finished') ? 'text-blue-500' : data.status.includes('Hiatus') ? 'text-yellow-500' : 'text-green-500'}`}
                >
                  {data.status}
                </span>
              </span>
              <span>Rank : {data.rank}</span>
              <span>Mean Score : {data.score}</span>
            </p>
          </div>
        </div>
      ) : (
        <>
          {err ? (
            <p>A ocurrido un error {err}</p>
          ) : (
            <div>
              <CircleLoading
                color="#536DFE"
                style={{ positionAbsolute: "none", margin: "auto" }}
              />
            </div>
          )}
        </>
      )}
    </>
  );
};


ManAnimeCard.propTypes = {
  manAnime : PropTypes.object.isRequired,
  context : PropTypes.string.isRequired,
  dispatchManime : PropTypes.func.isRequired
}