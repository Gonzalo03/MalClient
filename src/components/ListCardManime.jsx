import { ManAnimeCard } from "./ManAnimeCard";

export const ListCardManime = ({ manAnimes, context, dispatchManime }) => {
  return (
    <>
      <p className="text-center text-xl text-white">
        <span className="bg-primary p-2 rounded-xl">
          {context[0].toUpperCase() + context.substring(1)} Favorites
        </span>
      </p>
      <ul className="flex flex-wrap justify-center space-y-6 my-8">
        {manAnimes.map(
          (manAnime, index) =>
            index <= 1 && (
              <ManAnimeCard
                manAnime={manAnime}
                key={manAnime.mal_id}
                context={context}
                dispatchManime={dispatchManime}
              />
            )
        )}
      </ul>
    </>
  );
};
