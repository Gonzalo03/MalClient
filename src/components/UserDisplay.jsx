import { useRef, useEffect, useReducer } from "react";


import { useUser } from "../hook/useUser";
import { useManganimeMemo } from "../hook/useManganimeMemo";
import { avgReducer } from "../reducer/avgReducer";



import { CircleLoading } from "react-loadingg";
import { MenuStats } from "./MenuStats";
import { ListCardManime } from "./ListCardManime";

export const UserDisplay = ({ tag, deleteUser }) => {
  const { loading, data, err } = useUser(tag);

  const [avgManime, dispatchManime] = useReducer(avgReducer, {
    anime : [],
    manga : []
  });

  const computeAvgAnime = useManganimeMemo(avgManime.anime) || 0;
  const computeAvgManga = useManganimeMemo(avgManime.manga) || 0;

  const loadingRef = useRef();
  const aboutRef = useRef();

  useEffect(() => {
    loadingRef.current &&
      loadingRef.current.scrollIntoView({ behavior: "smooth" });

    if (aboutRef.current && data) {
      aboutRef.current.innerHTML = data.about || "";

      setTimeout(() => {
        aboutRef && aboutRef.current.scrollIntoView({ behavior: "smooth" });
      }, 100);
    }
  }, [tag, data]);



  return (
    <>
      {!loading && !err ? (
        <>
          <div className="md:grid grid-cols-2 my-5 userDisplay">
            {/* Profile Preview */}
            <div className="max-w-sm flex flex-col items-center justify-center rounded shadow-lg mx-auto">
              <img src={data.image_url} alt={tag} style={{ width: 300 }} />
              <div className="flex flex-col space-y-5 px-6 pt-4">
                <div className="font-bold text-xl mb-2">{tag}</div>
                <p
                  className="text-icon text-base overflow-hidden space-y-4"
                  ref={aboutRef}
                  style={{ width: 300 }}
                ></p>
                {computeAvgAnime &&<p
                    className={`text-${
                      computeAvgAnime > 7
                        ? "green"
                        : computeAvgAnime < 7 && computeAvgAnime > 4
                        ? "yellow"
                        : computeAvgAnime < 4
                        ? "red"
                        : ""
                    }-500`}
                  >
                    Anime score : {computeAvgAnime} ={">"}{" "}
                    {computeAvgAnime > 7
                      ? "GOD"
                      : computeAvgAnime < 7 && computeAvgAnime > 4
                      ? "Neutral"
                      : computeAvgAnime < 4
                      ? "ZZZZZZ"
                      : ""}
                  </p>}
                {computeAvgManga && (
                  <p
                    className={`text-${
                      computeAvgManga > 7
                        ? "green"
                        : computeAvgManga < 7 && computeAvgManga > 4
                        ? "yellow"
                        : computeAvgManga < 4
                        ? "red"
                        : ""
                    }-500`}
                  >
                    Manga score : {computeAvgManga} ={">"}{" "}
                    {computeAvgManga > 7
                      ? "GOD"
                      : computeAvgManga < 7 && computeAvgManga > 4
                      ? "Neutral"
                      : computeAvgManga < 4
                      ? "ZZZZZZ"
                      : ""}
                  </p>
                )}
              </div>
              <button
                className="bg-primary hover:bg-secondary text-white font-bold py-2 px-4 rounded-full my-5"
                onClick={() => deleteUser(tag)}
              >
                Remover
              </button>
            </div>

            {/* Stats Preview */}

            <div className="flex flex-col space-y-7 self-center mt-10">
              <MenuStats stats={data.anime_stats} />
              <MenuStats stats={data.manga_stats} />
            </div>
          </div>

          {/* Fvorites Preview */}

          <div className="mt-10">
            <ListCardManime
              manAnimes={data.favorites.anime}
              context="anime"
              dispatchManime={dispatchManime}
            />
            <ListCardManime
              manAnimes={data.favorites.manga}
              context="manga"
              dispatchManime={dispatchManime}
            />
          </div>
        </>
      ) : (
        <>
          {err ? (
            <p>A ocurrido un error {err}</p>
          ) : (
            <div ref={loadingRef}>
              <CircleLoading
                color="#536DFE"
                style={{ positionAbsolute: "none", margin: "auto" }}
              />
            </div>
          )}
        </>
      )}
    </>
  );
};
