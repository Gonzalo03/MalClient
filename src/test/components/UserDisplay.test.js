import { shallow } from 'enzyme'
import { UserDisplay } from '../../components/UserDisplay';
import { userFix } from '../___fixtures___/userFix';
import { useUser } from '../../hook/useUser';
jest.mock( '../../hook/useUser');


describe('Pruebas en el componente userDisplay', ()=>{


    const tag  = "MauroDT";

    const deleteUser = jest.fn();



    test('Debe hacer match con el snapshot', () => {
   
        useUser.mockReturnValue({
            lading : false,
            data : userFix, 
            err : null
        });
    
        const wrapper = shallow( <UserDisplay tag={tag} deleteUser={deleteUser} /> );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('MenuStats').length).toBe(2);

    });

    test('Debe hacer click', () => {
   
        useUser.mockReturnValue({
            lading : false,
            data : userFix, 
            err : null
        });
        
        const wrapper = shallow( <UserDisplay tag={tag} deleteUser={deleteUser} /> );

        wrapper.find('button').simulate('click');

        expect(deleteUser).toHaveBeenCalled();


    });


});




