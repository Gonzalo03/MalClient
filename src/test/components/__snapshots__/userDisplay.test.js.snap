// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente userDisplay Debe hacer match con el snapshot 1`] = `
<Fragment>
  <div
    className="md:grid grid-cols-2 my-5 userDisplay"
  >
    <div
      className="max-w-sm flex flex-col items-center justify-center rounded shadow-lg mx-auto"
    >
      <img
        alt="MauroDT"
        src="https://cdn.myanimelist.net/images/userimages/6484810.jpg?t=1633289400"
        style={
          Object {
            "width": 300,
          }
        }
      />
      <div
        className="flex flex-col space-y-5 px-6 pt-4"
      >
        <div
          className="font-bold text-xl mb-2"
        >
          MauroDT
        </div>
        <p
          className="text-icon text-base overflow-hidden space-y-4"
          style={
            Object {
              "width": 300,
            }
          }
        />
        0
        0
      </div>
      <button
        className="bg-primary hover:bg-secondary text-white font-bold py-2 px-4 rounded-full my-5"
        onClick={[Function]}
      >
        Remover
      </button>
    </div>
    <div
      className="flex flex-col space-y-7 self-center mt-10"
    >
      <MenuStats
        stats={
          Object {
            "completed": 142,
            "days_watched": 47.6,
            "dropped": 21,
            "episodes_watched": 2784,
            "mean_score": 6.11,
            "on_hold": 2,
            "plan_to_watch": 14,
            "rewatched": 0,
            "total_entries": 181,
            "watching": 2,
          }
        }
      />
      <MenuStats
        stats={
          Object {
            "chapters_read": 4315,
            "completed": 37,
            "days_read": 28.6,
            "dropped": 7,
            "mean_score": 6.78,
            "on_hold": 8,
            "plan_to_read": 14,
            "reading": 4,
            "reread": 1,
            "total_entries": 70,
            "volumes_read": 464,
          }
        }
      />
    </div>
  </div>
  <div
    className="mt-10"
  >
    <ListCardManime
      context="anime"
      dispatchManime={[Function]}
      manAnimes={
        Array [
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/4/5123.jpg?s=be2454544fb31044414aca031c5af1ed",
            "mal_id": 2001,
            "name": "Tengen Toppa Gurren Lagann",
            "url": "https://myanimelist.net/anime/2001/Tengen_Toppa_Gurren_Lagann",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/3/73178.jpg?s=80fcc4f3c1c109c4cb57426ce1370e84",
            "mal_id": 22535,
            "name": "Kiseijuu: Sei no Kakuritsu",
            "url": "https://myanimelist.net/anime/22535/Kiseijuu__Sei_no_Kakuritsu",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/5/12844.jpg?s=a17f80ba3bdc0e2bc634827a740700b0",
            "mal_id": 2164,
            "name": "Dennou Coil",
            "url": "https://myanimelist.net/anime/2164/Dennou_Coil",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/1295/106551.jpg?s=d80ac7dd1ba5b48d9ffb70b5c0723b4b",
            "mal_id": 37999,
            "name": "Kaguya-sama wa Kokurasetai: Tensai-tachi no Renai Zunousen",
            "url": "https://myanimelist.net/anime/37999/Kaguya-sama_wa_Kokurasetai__Tensai-tachi_no_Renai_Zunousen",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/1122/96435.jpg?s=ddb8a71f5d9f36507239e583e49b4d92",
            "mal_id": 28851,
            "name": "Koe no Katachi",
            "url": "https://myanimelist.net/anime/28851/Koe_no_Katachi",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/anime/3/72943.jpg?s=02a59f6ef79e706662dfc1e6f836f427",
            "mal_id": 28171,
            "name": "Shokugeki no Souma",
            "url": "https://myanimelist.net/anime/28171/Shokugeki_no_Souma",
          },
        ]
      }
    />
    <ListCardManime
      context="manga"
      dispatchManime={[Function]}
      manAnimes={
        Array [
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/3/179882.jpg?s=5c77b0059bc82990bab39002790802d4",
            "mal_id": 1706,
            "name": "JoJo no Kimyou na Bouken Part 7: Steel Ball Run",
            "url": "https://myanimelist.net/manga/1706/JoJo_no_Kimyou_na_Bouken_Part_7__Steel_Ball_Run",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/1/157897.jpg?s=8fb0808c48666aed744e94a2e67408c5",
            "mal_id": 2,
            "name": "Berserk",
            "url": "https://myanimelist.net/manga/2/Berserk",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/1/145619.jpg?s=5e2e7b1c23db307af7ffc87ee5e5040b",
            "mal_id": 768,
            "name": "Gunnm",
            "url": "https://myanimelist.net/manga/768/Gunnm",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/2/188925.jpg?s=cb8f03b751e591f609e9fd630c7e1a10",
            "mal_id": 642,
            "name": "Vinland Saga",
            "url": "https://myanimelist.net/manga/642/Vinland_Saga",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/2/25993.jpg?s=e5baecc03f66c3a52c9f5acab26f6d95",
            "mal_id": 7887,
            "name": "Toriko",
            "url": "https://myanimelist.net/manga/7887/Toriko",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/3/164420.jpg?s=71939fc4cb2146cabceb8275de5ce72e",
            "mal_id": 4632,
            "name": "Oyasumi Punpun",
            "url": "https://myanimelist.net/manga/4632/Oyasumi_Punpun",
          },
          Object {
            "image_url": "https://cdn.myanimelist.net/images/manga/3/188896.jpg?s=283305c99ce73633aeb4e53a766ba38c",
            "mal_id": 90125,
            "name": "Kaguya-sama wa Kokurasetai: Tensai-tachi no Renai Zunousen",
            "url": "https://myanimelist.net/manga/90125/Kaguya-sama_wa_Kokurasetai__Tensai-tachi_no_Renai_Zunousen",
          },
        ]
      }
    />
  </div>
</Fragment>
`;
